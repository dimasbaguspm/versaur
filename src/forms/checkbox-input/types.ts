import type {
  FieldsetHTMLAttributes,
  InputHTMLAttributes,
  ReactNode,
} from 'react'

/**
 * Props for the CheckboxInput component
 */
export interface CheckboxInputProps
  extends Omit<FieldsetHTMLAttributes<HTMLFieldSetElement>, 'children'> {
  /**
   * Label text to display above the checkbox group (rendered as legend)
   */
  label?: ReactNode
  /**
   * Helper text to display below the checkbox group
   */
  helperText?: ReactNode
  /**
   * Error message for invalid state
   */
  error?: ReactNode
  /**
   * Direction of checkbox layout
   */
  direction?: 'horizontal' | 'vertical'
  /**
   * Whether the checkbox group is required
   */
  required?: boolean
  /**
   * Children (CheckboxInput.Option components)
   */
  children: ReactNode
}

/**
 * Props for the CheckboxInput.Option component
 */
export interface CheckboxOptionProps
  extends Omit<InputHTMLAttributes<HTMLInputElement>, 'type'> {
  /**
   * The option label content
   */
  children: ReactNode
  /**
   * Description text below the option label
   */
  description?: ReactNode
}
