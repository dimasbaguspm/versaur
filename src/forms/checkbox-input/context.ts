import { createContext, useContext } from 'react'

interface CheckboxContextValue {
  disabled?: boolean
}

export const CheckboxContext = createContext<CheckboxContextValue | null>(null)

export const useCheckboxContext = () => {
  const context = useContext(CheckboxContext)
  if (!context) {
    throw new Error('CheckboxOption must be used within CheckboxInput')
  }
  return context
}
