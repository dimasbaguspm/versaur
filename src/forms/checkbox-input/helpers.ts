import { cva } from '@/utils/variants'

export const checkboxGroupVariants = cva('space-y-2', {
  variants: {
    direction: {
      vertical: 'space-y-2 space-x-0',
      horizontal: 'flex flex-wrap gap-4 space-y-0',
    },
  },
  defaultVariants: {
    direction: 'vertical',
  },
})

export const checkboxVariants = cva(
  'relative h-4 w-4 rounded border cursor-pointer transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed appearance-none bg-white border-primary/40 checked:bg-primary checked:border-primary focus:ring-primary/20'
)

export const checkboxLabelVariants = cva(
  'text-sm text-foreground cursor-pointer select-none',
  {
    variants: {
      disabled: {
        true: 'opacity-50 cursor-not-allowed',
        false: '',
      },
    },
    defaultVariants: {
      disabled: false,
    },
  }
)
