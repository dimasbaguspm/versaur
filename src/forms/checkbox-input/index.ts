export { CheckboxInput } from './checkbox-input'
export type { CheckboxInputProps, CheckboxOptionProps } from './types'
