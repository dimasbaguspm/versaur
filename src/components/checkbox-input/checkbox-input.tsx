import React from 'react'
import { cn } from '@/utils/cn'
import { checkboxGroupVariants } from './helpers'
import { CheckboxContext } from './context'
import { CheckboxOption } from './checkbox-input.atoms'
import type { CheckboxInputProps } from './types'

/**
 * CheckboxInput component for Versaur UI
 *
 * Provides a compound checkbox input with customizable styling and accessibility
 * Follows browser standards and supports multiple selection patterns
 * Uses custom ::after pseudo-element for checkmark styling instead of browser defaults
 */
const CheckboxInputRoot = React.forwardRef<HTMLDivElement, CheckboxInputProps>(
  (
    {
      variant = 'primary',
      size = 'md',
      label,
      helperText,
      error,
      direction = 'vertical',
      className,
      disabled,
      children,
      ...props
    },
    ref
  ) => {
    const hasError = Boolean(error)

    const contextValue = {
      variant,
      size,
      disabled,
      error: hasError,
    }

    return (
      <CheckboxContext.Provider value={contextValue}>
        <div ref={ref} className={cn('w-full', className)} {...props}>
          {label && (
            <div className='block text-sm font-medium text-foreground mb-3'>
              {label}
            </div>
          )}
          <div className={checkboxGroupVariants({ direction })}>{children}</div>
          {hasError && (
            <div className='mt-2 text-sm text-danger' role='alert'>
              {error}
            </div>
          )}
          {!hasError && helperText && (
            <div className='mt-2 text-sm text-gray-600'>{helperText}</div>
          )}
        </div>
      </CheckboxContext.Provider>
    )
  }
)

CheckboxInputRoot.displayName = 'CheckboxInput'

export const CheckboxInput = Object.assign(CheckboxInputRoot, {
  Option: CheckboxOption,
})
