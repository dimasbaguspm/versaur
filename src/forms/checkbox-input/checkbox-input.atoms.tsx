import React from 'react'
import { cn } from '@/utils/cn'
import { checkboxVariants, checkboxLabelVariants } from './helpers'
import { useCheckboxContext } from './context'
import type { CheckboxOptionProps } from './types'

/**
 * CheckboxOption atom component
 *
 * Individual checkbox option that can be used within CheckboxInput
 * Features custom styling with ::after pseudo-element for checkmark
 */
export const CheckboxOption = React.forwardRef<
  HTMLInputElement,
  CheckboxOptionProps
>(({ children, description, className, disabled, id, ...props }, ref) => {
  const context = useCheckboxContext()
  const generatedId = React.useId()
  const inputId = id || generatedId

  const isDisabled = disabled || context.disabled

  return (
    <div className='flex items-start gap-2'>
      <div className='relative flex items-center mt-0.5'>
        <input
          ref={ref}
          type='checkbox'
          id={inputId}
          disabled={isDisabled}
          className={cn(
            checkboxVariants(),
            // Custom checkmark styling using ::after - centered positioning
            'after:content-[""] after:absolute after:left-1/2 after:top-[45%] after:-translate-x-1/2 after:-translate-y-1/2 after:border-b-2 after:border-r-2 after:border-white after:rotate-45 after:opacity-0 after:transition-opacity after:duration-200',
            // Checkmark dimensions for md size (default)
            'after:w-[4px] after:h-[8px]',
            'checked:after:opacity-100',
            className
          )}
          {...props}
        />
      </div>
      <div className='flex flex-col'>
        <label
          htmlFor={inputId}
          className={checkboxLabelVariants({
            disabled: isDisabled,
          })}
        >
          {children}
        </label>
        {description && (
          <div
            className={cn(
              'text-xs text-gray-600 mt-1',
              isDisabled && 'opacity-50'
            )}
          >
            {description}
          </div>
        )}
      </div>
    </div>
  )
})

CheckboxOption.displayName = 'CheckboxOption'
