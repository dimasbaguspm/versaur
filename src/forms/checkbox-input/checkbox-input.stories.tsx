import type { Meta, StoryObj } from '@storybook/react'
import { CheckboxInput } from './checkbox-input'

/**
 * CheckboxInput is a compound component that provides checkbox inputs with primary variant styling
 * Uses fieldset/legend for proper semantic structure and accessibility
 *
 * Key features:
 * - Custom styled checkboxes using ::after pseudo-elements
 * - Primary variant using Versaur's coral color
 * - Flexible compound pattern for composition
 * - Full accessibility support with proper ARIA attributes and fieldset/legend structure
 * - Support for both horizontal and vertical layouts
 * - Required field indicator with asterisk
 * - Error state that only affects error text, not checkbox styling
 */
const meta: Meta<typeof CheckboxInput> = {
  title: 'Forms/CheckboxInput',
  component: CheckboxInput,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'A flexible checkbox input component supporting multiple selection with primary variant styling and proper semantic structure using fieldset/legend.',
      },
    },
  },
  argTypes: {
    direction: {
      control: 'select',
      options: ['horizontal', 'vertical'],
    },
  },
}

export default meta
type Story = StoryObj<typeof CheckboxInput>

/**
 * Default checkbox input with multiple options
 */
export const Default: Story = {
  args: {
    label: 'Choose your preferences',
    direction: 'vertical',
  },
  render: args => (
    <CheckboxInput {...args}>
      <CheckboxInput.Option value='option1'>
        Newsletter subscription
      </CheckboxInput.Option>
      <CheckboxInput.Option value='option2'>
        Marketing emails
      </CheckboxInput.Option>
      <CheckboxInput.Option value='option3'>
        Product updates
      </CheckboxInput.Option>
    </CheckboxInput>
  ),
}

/**
 * Checkbox options with descriptions and required indicator
 */
export const WithDescriptions: Story = {
  args: {
    label: 'Privacy settings',
    required: true,
  },
  render: args => (
    <CheckboxInput {...args}>
      <CheckboxInput.Option
        value='analytics'
        description='Help us improve by sharing anonymous usage data'
      >
        Analytics tracking
      </CheckboxInput.Option>
      <CheckboxInput.Option
        value='marketing'
        description='Receive personalized offers and recommendations'
      >
        Marketing communications
      </CheckboxInput.Option>
      <CheckboxInput.Option
        value='essential'
        description='Required for core functionality and security'
        defaultChecked
        disabled
      >
        Essential cookies
      </CheckboxInput.Option>
    </CheckboxInput>
  ),
}

/**
 * Horizontal layout for compact displays
 */
export const HorizontalLayout: Story = {
  args: {
    label: 'Skills',
    direction: 'horizontal',
  },
  render: args => (
    <CheckboxInput {...args}>
      <CheckboxInput.Option value='react'>React</CheckboxInput.Option>
      <CheckboxInput.Option value='typescript'>TypeScript</CheckboxInput.Option>
      <CheckboxInput.Option value='nodejs'>Node.js</CheckboxInput.Option>
      <CheckboxInput.Option value='python'>Python</CheckboxInput.Option>
    </CheckboxInput>
  ),
}

/**
 * Error state demonstration - error only affects error text, not checkbox styling
 */
export const WithError: Story = {
  args: {
    label: 'Terms and conditions',
    required: true,
    error: 'You must accept the terms to continue',
  },
  render: args => (
    <CheckboxInput {...args}>
      <CheckboxInput.Option value='terms'>
        I accept the terms and conditions
      </CheckboxInput.Option>
      <CheckboxInput.Option value='privacy'>
        I accept the privacy policy
      </CheckboxInput.Option>
    </CheckboxInput>
  ),
}

/**
 * Disabled state demonstration
 */
export const Disabled: Story = {
  args: {
    label: 'Disabled options',
    disabled: true,
    helperText: 'These options are currently unavailable',
  },
  render: args => (
    <CheckboxInput {...args}>
      <CheckboxInput.Option value='disabled1'>
        Disabled option 1
      </CheckboxInput.Option>
      <CheckboxInput.Option value='disabled2'>
        Disabled option 2
      </CheckboxInput.Option>
    </CheckboxInput>
  ),
}
