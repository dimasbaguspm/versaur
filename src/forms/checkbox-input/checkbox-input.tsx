import React from 'react'
import { cn } from '@/utils/cn'
import { checkboxGroupVariants } from './helpers'
import { CheckboxContext } from './context'
import { CheckboxOption } from './checkbox-input.atoms'
import type { CheckboxInputProps } from './types'

/**
 * CheckboxInput component for Versaur UI
 *
 * Provides a compound checkbox input with customizable styling and accessibility
 * Follows browser standards and supports multiple selection patterns
 * Uses fieldset/legend structure for proper semantics
 * Uses custom ::after pseudo-element for checkmark styling instead of browser defaults
 */
const CheckboxInputRoot = React.forwardRef<
  HTMLFieldSetElement,
  CheckboxInputProps
>(
  (
    {
      label,
      helperText,
      error,
      direction = 'vertical',
      className,
      disabled,
      required,
      children,
      ...props
    },
    ref
  ) => {
    const hasError = Boolean(error)

    const contextValue = {
      disabled,
    }

    return (
      <CheckboxContext.Provider value={contextValue}>
        <fieldset
          ref={ref}
          disabled={disabled}
          className={cn('w-full border-0 p-0 m-0', className)}
          {...props}
        >
          {label && (
            <legend className='block text-sm font-medium text-foreground mb-3'>
              {label}
              {required && (
                <span className='text-danger ml-1' aria-label='required'>
                  *
                </span>
              )}
            </legend>
          )}
          <div className={checkboxGroupVariants({ direction })}>{children}</div>
          {hasError && (
            <div className='mt-2 text-sm text-danger' role='alert'>
              {error}
            </div>
          )}
          {!hasError && helperText && (
            <div className='mt-2 text-sm text-gray-600'>{helperText}</div>
          )}
        </fieldset>
      </CheckboxContext.Provider>
    )
  }
)

CheckboxInputRoot.displayName = 'CheckboxInput'

export const CheckboxInput = Object.assign(CheckboxInputRoot, {
  Option: CheckboxOption,
})
