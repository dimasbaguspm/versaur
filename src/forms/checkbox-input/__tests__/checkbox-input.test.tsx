import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { composeStories } from '@storybook/react'
import { describe, expect, it } from 'vitest'
import * as stories from '../checkbox-input.stories'

const { Default, WithDescriptions, WithError, Disabled } =
  composeStories(stories)

describe('CheckboxInput', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Default />)
    expect(asFragment()).toMatchSnapshot()
  })

  describe('Basic functionality', () => {
    it('renders checkbox options correctly with fieldset/legend', () => {
      render(<Default />)

      const fieldset = screen.getByRole('group')
      expect(fieldset).toBeInTheDocument()
      expect(fieldset.tagName).toBe('FIELDSET')

      expect(screen.getByText('Choose your preferences')).toBeInTheDocument()
      expect(
        screen.getByLabelText('Newsletter subscription')
      ).toBeInTheDocument()
      expect(screen.getByLabelText('Marketing emails')).toBeInTheDocument()
      expect(screen.getByLabelText('Product updates')).toBeInTheDocument()
    })

    it('allows checking and unchecking options', async () => {
      const user = userEvent.setup()
      render(<Default />)

      const newsletter = screen.getByLabelText('Newsletter subscription')
      const marketing = screen.getByLabelText('Marketing emails')

      expect(newsletter).not.toBeChecked()
      expect(marketing).not.toBeChecked()

      await user.click(newsletter)
      expect(newsletter).toBeChecked()

      await user.click(marketing)
      expect(marketing).toBeChecked()

      await user.click(newsletter)
      expect(newsletter).not.toBeChecked()
      expect(marketing).toBeChecked()
    })
  })

  describe('With descriptions and required indicator', () => {
    it('renders option descriptions', () => {
      render(<WithDescriptions />)

      expect(
        screen.getByText('Help us improve by sharing anonymous usage data')
      ).toBeInTheDocument()
      expect(
        screen.getByText('Receive personalized offers and recommendations')
      ).toBeInTheDocument()
      expect(
        screen.getByText('Required for core functionality and security')
      ).toBeInTheDocument()
    })

    it('displays required asterisk', () => {
      render(<WithDescriptions />)

      const asterisk = screen.getByLabelText('required')
      expect(asterisk).toBeInTheDocument()
      expect(asterisk).toHaveClass('text-danger')
    })

    it('handles disabled options correctly', () => {
      render(<WithDescriptions />)

      const essentialCookies = screen.getByLabelText('Essential cookies')
      expect(essentialCookies).toBeDisabled()
      expect(essentialCookies).toBeChecked()
    })
  })

  describe('Error states', () => {
    it('displays error message', () => {
      render(<WithError />)

      const errorMessage = screen.getByRole('alert')
      expect(errorMessage).toHaveTextContent(
        'You must accept the terms to continue'
      )
    })

    it('error does not affect checkbox styling, only error text', () => {
      render(<WithError />)

      const checkbox = screen.getByLabelText(
        'I accept the terms and conditions'
      )
      // Checkbox should still have primary styling
      expect(checkbox).toHaveClass(/border-primary/)
      expect(checkbox).not.toHaveClass(/danger/)

      // Error text should be displayed
      expect(screen.getByRole('alert')).toHaveClass('text-danger')
    })
  })

  describe('Disabled state', () => {
    it('disables all options when fieldset is disabled', () => {
      render(<Disabled />)

      const option1 = screen.getByLabelText('Disabled option 1')
      const option2 = screen.getByLabelText('Disabled option 2')

      expect(option1).toBeDisabled()
      expect(option2).toBeDisabled()

      // Fieldset should be disabled
      const fieldset = screen.getByRole('group')
      expect(fieldset).toBeDisabled()
    })

    it('shows helper text for disabled state', () => {
      render(<Disabled />)

      expect(
        screen.getByText('These options are currently unavailable')
      ).toBeInTheDocument()
    })
  })

  describe('Accessibility', () => {
    it('has proper ARIA attributes', () => {
      render(<Default />)

      const checkboxes = screen.getAllByRole('checkbox')
      checkboxes.forEach(checkbox => {
        expect(checkbox).toHaveAttribute('type', 'checkbox')
      })
    })

    it('uses fieldset and legend for proper semantics', () => {
      render(<Default />)

      const fieldset = screen.getByRole('group')
      expect(fieldset.tagName).toBe('FIELDSET')

      const legend = fieldset.querySelector('legend')
      expect(legend).toBeInTheDocument()
      expect(legend).toHaveTextContent('Choose your preferences')
    })

    it('associates labels with checkboxes correctly', () => {
      render(<Default />)

      const newsletter = screen.getByLabelText('Newsletter subscription')
      const marketing = screen.getByLabelText('Marketing emails')

      expect(newsletter).toBeInstanceOf(HTMLInputElement)
      expect(marketing).toBeInstanceOf(HTMLInputElement)
    })

    it('supports keyboard navigation', async () => {
      const user = userEvent.setup()
      render(<Default />)

      const newsletter = screen.getByLabelText('Newsletter subscription')

      await user.tab()
      expect(newsletter).toHaveFocus()

      await user.keyboard('[Space]')
      expect(newsletter).toBeChecked()
    })
  })
})
