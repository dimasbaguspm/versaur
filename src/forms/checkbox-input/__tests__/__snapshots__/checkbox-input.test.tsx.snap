// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CheckboxInput > should match snapshot 1`] = `
<DocumentFragment>
  <fieldset
    class="w-full border-0 p-0 m-0"
  >
    <legend
      class="block text-sm font-medium text-foreground mb-3"
    >
      Choose your preferences
    </legend>
    <div
      class="space-y-2 space-y-2 space-x-0"
    >
      <div
        class="flex items-start gap-2"
      >
        <div
          class="relative flex items-center mt-0.5"
        >
          <input
            class="relative h-4 w-4 rounded border cursor-pointer transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed appearance-none bg-white border-primary/40 checked:bg-primary checked:border-primary focus:ring-primary/20 after:content-[""] after:absolute after:left-1/2 after:top-[45%] after:-translate-x-1/2 after:-translate-y-1/2 after:border-b-2 after:border-r-2 after:border-white after:rotate-45 after:opacity-0 after:transition-opacity after:duration-200 after:w-[4px] after:h-[8px] checked:after:opacity-100"
            id="«r0»"
            type="checkbox"
            value="option1"
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-sm text-foreground cursor-pointer select-none"
            for="«r0»"
          >
            Newsletter subscription
          </label>
        </div>
      </div>
      <div
        class="flex items-start gap-2"
      >
        <div
          class="relative flex items-center mt-0.5"
        >
          <input
            class="relative h-4 w-4 rounded border cursor-pointer transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed appearance-none bg-white border-primary/40 checked:bg-primary checked:border-primary focus:ring-primary/20 after:content-[""] after:absolute after:left-1/2 after:top-[45%] after:-translate-x-1/2 after:-translate-y-1/2 after:border-b-2 after:border-r-2 after:border-white after:rotate-45 after:opacity-0 after:transition-opacity after:duration-200 after:w-[4px] after:h-[8px] checked:after:opacity-100"
            id="«r1»"
            type="checkbox"
            value="option2"
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-sm text-foreground cursor-pointer select-none"
            for="«r1»"
          >
            Marketing emails
          </label>
        </div>
      </div>
      <div
        class="flex items-start gap-2"
      >
        <div
          class="relative flex items-center mt-0.5"
        >
          <input
            class="relative h-4 w-4 rounded border cursor-pointer transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed appearance-none bg-white border-primary/40 checked:bg-primary checked:border-primary focus:ring-primary/20 after:content-[""] after:absolute after:left-1/2 after:top-[45%] after:-translate-x-1/2 after:-translate-y-1/2 after:border-b-2 after:border-r-2 after:border-white after:rotate-45 after:opacity-0 after:transition-opacity after:duration-200 after:w-[4px] after:h-[8px] checked:after:opacity-100"
            id="«r2»"
            type="checkbox"
            value="option3"
          />
        </div>
        <div
          class="flex flex-col"
        >
          <label
            class="text-sm text-foreground cursor-pointer select-none"
            for="«r2»"
          >
            Product updates
          </label>
        </div>
      </div>
    </div>
  </fieldset>
</DocumentFragment>
`;
